File: compare/golang/graph/graph.go
package graph

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"

	"github.com/timHau/extensor_coding/extensor"
	"github.com/timHau/extensor_coding/matrix"
)

type Graph struct {
	AdjMat *matrix.Matrix
}

func createBernoulli(n int, k int) []*extensor.Extensor {
	res := make([]*extensor.Extensor, n)

	for i := 0; i < n; i++ {
		coeffs := make([]int, k)
		basis := [][]uint8{}
		for j := 1; j <= k; j++ {
			r := rand.Intn(2)
			if r == 0 {
				r = -1
			}
			coeffs[j-1] = r
			basis = append(basis, []uint8{uint8(j)})
		}
		e := extensor.New(coeffs, basis)
		res[i] = e.Lift(k)
	}

	return res
}

func AdjMatFromTsv(path string) (int, int, []uint8) {
	file, err := os.Open(path)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	trimmed := strings.Replace(lines[1], "% ", "", -1)
	splited := strings.Split(trimmed, " ")

	nrows, err := strconv.Atoi(splited[1])
	if err != nil {
		panic(err)
	}

	ncols, err := strconv.Atoi(splited[2])
	if err != nil {
		panic(err)
	}

	adjMat := make([]uint8, nrows*ncols)
	for _, line := range lines {
		if !strings.HasPrefix(line, "%") {
			values := strings.Split(line, " ")

			from, err := strconv.Atoi(values[0])
			if err != nil {
				panic(err)
			}

			to, err := strconv.Atoi(values[1])
			if err != nil {
				panic(err)
			}

			from = from - 1
			to = to - 1
			adjMat[from*ncols+to] = 1
		}
	}

	return nrows, ncols, adjMat
}

func NewWithCoding(k int, nrows int, ncols int, data []uint8) (*Graph, []extensor.Extensor) {
	coding := createBernoulli(ncols, k)

	adjMat := make([]*extensor.Extensor, len(data))
	for i, v := range data {
		if v != 0 {
			rowIndex := i / ncols
			adjMat[i] = coding[rowIndex]
		}
	}

	coding_into := make([]extensor.Extensor, len(coding))
	for i, v := range coding {
		coding_into[i] = *v
	}

	return &Graph{
		AdjMat: matrix.New(nrows, ncols, adjMat),
	}, coding_into
}

func (g *Graph) ComputeWalkSum(k int, coding []extensor.Extensor) int {
	b := g.AdjMat.Mul(coding)
	for i := 1; i < k-1; i++ {
		b = g.AdjMat.Mul(b)
	}

	fmt.Println(coding)

	resExt := extensor.Zero()
	for _, e := range b {
		resExt = resExt.Add(e)
	}

	res := 0
	for _, c := range resExt.Coeffs() {
		res += c
	}

	return res
}

File: compare/cpp/graph.cpp
#include "graph.h"

std::tuple<Graph, vector<Extensor>> Graph::from_tsv_with_coding(string path_str, int k)
{
	std::ifstream file(path_str);
	string line;

	vector<string> lines;
	while (std::getline(file, line)) {
		lines.push_back(line);
	}

	std::istringstream iss(lines[1]);
	vector<string> dim{std::istream_iterator<string>(iss), {}};
	int nrows = std::stoi(dim[2]);
	int ncols = std::stoi(dim[3]);

	auto coding = create_bernoulli(nrows + ncols, k);

	vector<Extensor> adj_mat_data(nrows * ncols, Extensor::zero());
	for (const auto& line : lines) {
		if (line.rfind("%", 0) != 0) { // lines does not start with %
			std::istringstream is(line);
			vector<string> values{std::istream_iterator<string>(is), {}}; // split whitespace

			int from = std::stoi(values[0]) - 1;
			int to = std::stoi(values[1]) - 1;
			adj_mat_data[from * ncols + to] = coding[from];
		}
	}

	Matrix adj_mat = Matrix(nrows, ncols, adj_mat_data);
	Graph g{adj_mat, nrows + ncols};
	return std::make_tuple(g, coding);
}

vector<Extensor> Graph::create_bernoulli(int n, int k)
{
	vector<Extensor> res(n, Extensor::zero());

	srand(time(NULL)); // init random seed
	for (int i = 0; i < n; ++i) {
		vector<int> coeffs(k);
		vector<vector<int>> base(k);
		for (int j = 0; j < k; ++j) {
			int r = rand() % 2;
			if (r == 0) {
				r = -1;
			}
			coeffs[j] = r;
			base[j] = vector<int>{j+1};
		}
		Extensor e{coeffs, base};
		res[i] = e.lift(k);
	}

	return res;
}

Extensor Graph::compute_walk_sum(int k, vector<Extensor> coding)
{
	vector<Extensor> b = this->adj_mat * coding;

	for (int i = 1; i < k-1; ++i) {
		b = this->adj_mat * b;
	}

	Extensor res = Extensor::zero();
	for (const auto& e : b) {
		res = res + e;
	}

	return res;
}